File: ditto-nodejs-client/src/client-internal.ts
import { DittoError } from './types.js';

const STRICT_TOKEN_RE = /^[A-Za-z0-9._:-]+$/;
const STRICT_PATTERN_RE = /^[A-Za-z0-9._:\-*]+$/;

export type StrictCoreOperation =
  | 'delete'
  | 'get'
  | 'incr'
  | 'set'
  | 'setNX'
  | 'unwatch'
  | 'watch';

export type StrictPatternOperation = 'deleteByPattern' | 'setTtlByPattern';

export function validateCoreInputs(
  strictMode: boolean,
  op: StrictCoreOperation,
  key: string,
  namespace?: string,
): void {
  if (!strictMode) return;
  const keyTrimmed = key.trim();
  if (keyTrimmed.length === 0) {
    throw new Error(`Invalid ${op} request: key must not be empty.`);
  }
  if (!STRICT_TOKEN_RE.test(key)) {
    throw new Error(
      `Invalid ${op} request: key contains unsupported characters. Allowed: [A-Za-z0-9._:-]`,
    );
  }
  validateNamespace(op, namespace);
}

export function validatePatternInputs(
  strictMode: boolean,
  op: StrictPatternOperation,
  pattern: string,
  namespace?: string,
): void {
  if (!strictMode) return;
  const patternTrimmed = pattern.trim();
  if (patternTrimmed.length === 0) {
    throw new Error(`Invalid ${op} request: pattern must not be empty.`);
  }
  if (!STRICT_PATTERN_RE.test(patternTrimmed)) {
    throw new Error(
      `Invalid ${op} request: pattern contains unsupported characters. Allowed: [A-Za-z0-9._:-*]`,
    );
  }
  validateNamespace(op, namespace);
}

export function normalizeNamespace(namespace?: string): string | undefined {
  const trimmed = namespace?.trim();
  return trimmed && trimmed.length > 0 ? trimmed : undefined;
}

export function watchCallbackKey(key: string, namespace?: string): string {
  return namespace ? `${namespace}::${key}` : key;
}

export async function unsupportedAtomicHttpError(
  resp: Response,
  operation: 'SET_NX' | 'INCR',
): Promise<Error> {
  try {
    const body = await resp.json() as { error?: string; message?: string };
    if (body.error === 'UnsupportedRequest') {
      return new DittoError('UnsupportedRequest', body.message ?? 'UnsupportedRequest');
    }
  } catch {
    // fall through to normalized message below
  }
  return unsupportedAtomicOperationError(operation);
}

export function normalizeAtomicUnsupportedError(
  error: unknown,
  operation: 'SET_NX' | 'INCR',
): Error {
  if (error instanceof DittoError) {
    return error;
  }
  const message = error instanceof Error ? error.message : String(error);
  const normalized = message.toLowerCase();
  if (
    normalized.includes('unsupported')
    || normalized.includes('protocol')
    || normalized.includes('decode')
    || normalized.includes('missing client_response')
    || normalized.includes('oneof has no active field')
    || normalized.includes('unexpected response')
    || normalized.includes('socket hang up')
    || normalized.includes('econnreset')
    || normalized.includes('unexpected eof')
  ) {
    return unsupportedAtomicOperationError(operation);
  }
  return error instanceof Error ? error : new Error(message);
}

function unsupportedAtomicOperationError(operation: 'SET_NX' | 'INCR'): DittoError {
  return new DittoError(
    'UnsupportedRequest',
    `UnsupportedRequest: server does not support ${operation}. Upgrade dittod to a version with atomic primitives.`,
  );
}

function validateNamespace(op: StrictCoreOperation | StrictPatternOperation, namespace?: string): void {
  if (namespace === undefined) return;
  const nsTrimmed = namespace.trim();
  if (nsTrimmed.length === 0) {
    throw new Error(`Invalid ${op} request: namespace must not be blank when provided.`);
  }
  if (nsTrimmed.includes('::')) {
    throw new Error(`Invalid ${op} request: namespace must not contain '::'.`);
  }
  if (!STRICT_TOKEN_RE.test(nsTrimmed)) {
    throw new Error(
      `Invalid ${op} request: namespace contains unsupported characters. Allowed: [A-Za-z0-9._:-]`,
    );
  }
}

File: ditto-go-client/client_internal.go
package ditto

import (
	"encoding/json"
	"fmt"
	"strings"
)

func resolveCoreNamespace(strict bool, op string, key string, namespace ...string) (*string, error) {
	ns, err := normalizedNamespaceStrict(strict, namespace...)
	if err != nil {
		return nil, err
	}
	if err := validateCoreInputs(strict, op, key, ns); err != nil {
		return nil, err
	}
	return ns, nil
}

func resolveCoreNamespaceValue(strict bool, op string, key string, namespace string) (*string, error) {
	return resolveCoreNamespace(strict, op, key, namespace)
}

func resolvePatternNamespace(strict bool, op string, pattern string, namespace ...string) (*string, error) {
	ns, err := normalizedNamespaceStrict(strict, namespace...)
	if err != nil {
		return nil, err
	}
	if err := validatePatternInputs(strict, op, pattern, ns); err != nil {
		return nil, err
	}
	return ns, nil
}

func optionalUint64FromSlice(values ...uint64) *uint64 {
	if len(values) == 0 || values[0] == 0 {
		return nil
	}
	return &values[0]
}

func optionalUint64(value uint64) *uint64 {
	if value == 0 {
		return nil
	}
	return &value
}

func optionalInt64(value int64) *int64 {
	return &value
}

func normalizeNamespace(namespace ...string) *string {
	ns, _ := normalizedNamespaceStrict(false, namespace...)
	return ns
}

func parseUnexpectedResponseError() error {
	return fmt.Errorf("unexpected response")
}

func unsupportedAtomicOperationError(operation string) error {
	return &DittoError{
		Code:    ErrUnsupportedRequest,
		Message: fmt.Sprintf("server does not support %s; upgrade dittod to a version with atomic primitives", operation),
	}
}

func parseAtomicHTTPUnsupported(body []byte, operation string) error {
	var payload struct {
		Error   string `json:"error"`
		Message string `json:"message"`
	}
	if json.Unmarshal(body, &payload) == nil && payload.Error == ErrUnsupportedRequest {
		msg := payload.Message
		if msg == "" {
			msg = payload.Error
		}
		return &DittoError{Code: ErrUnsupportedRequest, Message: msg}
	}
	return unsupportedAtomicOperationError(operation)
}

func normalizeAtomicTCPErr(err error, operation string) error {
	if err == nil {
		return nil
	}
	if de, ok := err.(*DittoError); ok {
		return de
	}
	msg := strings.ToLower(err.Error())
	if strings.Contains(msg, "unsupported") ||
		strings.Contains(msg, "protocol") ||
		strings.Contains(msg, "decode") ||
		strings.Contains(msg, "clientresponse oneof") ||
		strings.Contains(msg, "unexpected response") ||
		strings.Contains(msg, "eof") ||
		strings.Contains(msg, "connection reset") {
		return unsupportedAtomicOperationError(operation)
	}
	return err
}
